sw/host/opentitanlib/src/rom_patch/mod.rs

```diff
@@ -0,0 +1,691 @@
+    match_base: u32,
+    remap_base: u32,
+}
+
+// Patch locked field position in the match descriptor.
+const PATCH_LOCKED_FIELD_SHIFT: usize = 31;
+
+// Patch size field position in the match descriptor.
+const PATCH_SIZE_FIELD_SHIFT: usize = 27;
+const PATCH_SIZE_FIELD_WIDTH: usize = 4;
+const PATCH_SIZE_FIELD_MASK: u32 = (1 << PATCH_SIZE_FIELD_WIDTH) - 1;
+
+// Patch match address in the match descriptor.
+const PATCH_MATCH_ADDR_FIELD_WIDTH: usize = 27;
+const PATCH_MATCH_ADDR_FIELD_MASK: u32 = (1 << PATCH_MATCH_ADDR_FIELD_WIDTH) - 1;
+
+impl RomPatchMatchDescriptor {
+    fn from_hjson(remap_base: u32, region: &RomPatchRegionHjson) -> Result<Self> {
+        let match_addr = region.match_addr;
+        let patch_code_size_words = region.patch_code.len();
+
+        let size_dwords = match patch_code_size_words {
+            1 | 2 | 4 | 8 => (patch_code_size_words as u32) & PATCH_SIZE_FIELD_MASK,
+            _ => bail!("Invalid ROM patch region code size {patch_code_size_words}"),
+        };
+        let locked = if region.locked {
+            1 << PATCH_LOCKED_FIELD_SHIFT
+        } else {
+            0
+        };
+
+        let match_address = match match_addr {
+            addr if addr <= PATCH_MATCH_ADDR_FIELD_MASK => addr & PATCH_MATCH_ADDR_FIELD_MASK,
+            _ => bail!("Invalid ROM patch match address {match_addr}"),
+        };
+
+        let match_base = locked | (size_dwords << PATCH_SIZE_FIELD_SHIFT) | match_address;
+
+        Ok(RomPatchMatchDescriptor {
+            match_base,
+            remap_base,
+        })
+    }
+}
+
+#[derive(AsBytes, Copy, Clone, Debug, Default)]
+#[repr(packed, C)]
+struct RomPatchMatchTable {
+    table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS],
+}
+
+impl RomPatchMatchTable {
+    fn from_hjson(base_addr: u32, regions: &[RomPatchRegionHjson]) -> Result<Self> {
+        let mut table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS] = Default::default();
+        let mut remap_addr = base_addr;
+        for (index, region) in regions.iter().enumerate() {
+            let match_pair = RomPatchMatchDescriptor::from_hjson(remap_addr, region)?;
+            table[index] = match_pair;
+            remap_addr += (region.patch_code.len() as u32) * mem::size_of::<u32>() as u32;
+        }
+
+        Ok(Self { table })
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct RomPatchCode {
+    code: Vec<u32>,
+}
+
+impl RomPatchCode {
+    fn from_hjson(regions: &Vec<RomPatchRegionHjson>) -> Result<Self> {
+        let mut code_size_words = regions.iter().map(|r| r.patch_code.len()).sum();
+
+        // Compute the overall code size for the patch.
+        for region in regions {
+            if let Some(sub_routines_hjson) = &region.sub_routines {
+                for s in sub_routines_hjson {
+                    code_size_words =
+                        std::cmp::max(code_size_words, (s.offset >> 2) + s.code.len());
+                }
+            }
+        }
```

Comment:
You could remove the Option from `sub_routines` combined with `#[serde(default)]` to make it an empty vector in case it is not defined in the hjson, so that the `if let` would not be required. 
Iterator functions could also be used to make the code more idiomatic.
```suggestion
let code_size_words = regions
        .iter()
        .flat_map(|region| region.sub_routines.iter())
        .map(|s| (s.offset >> 2) + s.code.len())
        .fold(code_size_words, std::cmp::max);
```

Reply:
Nice. Applied, thanks.